makeModulesClosure: fixup firmware extraction

baloo opened the pull request:
###### Motivation for this change
When building on a non nixos platform, I would get error trying to lookup the firmware for unix module. `modinfo -F firmware unix` output looks wrong. This is an attempt  to work around that. Also I'm not sure we were loading module information from the to-be-loaded kernel, but rather the currently running one.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

baloo committed:
makeModulesClosure: fixup firmware extraction

After a recent upgrade of modinfo, its output is now incorrect for
builtin modules. This commit filters out the output until a fix is made
available upstream

baloo commented on pkgs/build-support/kernel/modules-closure.sh:
```diff
@@ -68,7 +68,11 @@ done
+    for i in $(modinfo -b $kernel --set-version "$version" -F firmware $module); do
+        # $ modinfo -F firmware unix
+        # name:           unix
+        if test "$i" = "name:"; then break; fi
```
I somewhat agree `modinfo` acts weird, piping it to `| grep -v` actually sounds better. Let me know.

lheckemann replied:
Yeah, while the kmod fix is my main preference, my preferred _workaround_ is the grep approach, since otherwise we'll fail to load firmware for any built-in modules that require it.

xaverdh commented:
Ok, the following patch works for me: [0001-modinfo-dont-print-module-name-when-field-is-given.txt](https://github.com/NixOS/nixpkgs/files/5115021/0001-modinfo-dont-print-module-name-when-field-is-given.txt)
Is linux-modules@vger.kernel.org the right place to post this?

baloo commented:
To note: There was also fixes on the kernel modules to read included. If #96123 is merged, we still need those.

baloo commented on pkgs/build-support/kernel/modules-closure.sh:
```diff
@@ -68,7 +68,11 @@ done
 for module in $(cat closure); do
-    for i in $(modinfo -F firmware $module); do
+    for i in $(modinfo -b $kernel --set-version "$version" -F firmware $module); do
+        # $ modinfo -F firmware unix
+        # name:           unix
+        if test "$i" = "name:"; then break; fi
```
Once modinfo patches are merged, we still need this in:
`-b $kernel --set-version "$version" `
```suggestion
```

xaverdh commented:
Since the patch for `kmod` is a mass rebuild and did not make it into the last staging round before branch-off for 20.09, something like this should probably go into master (before branch-off).

baloo committed:
makeModulesClosuse: read modules from corrent kernel version

Before this commit, the firmware information would be loaded from the
currently running kernel, not from the kernel to be loaded.

This commit ensures the correct kernel version and modules are read.

baloo commented:
Splitted the two commits, and rewrote the fix to use a `| grep -v` instead as discussed.
